WordPress/src/test/java/org/wordpress/android/ui/mysite/cards/dashboard/CardsSourceTest.kt

```diff
@@ -0,0 +1,234 @@
+
+        assertThat(result.size).isEqualTo(1)
+        assertThat(result.first()).isEqualTo(CardsUpdate(data))
+    }
+
+    @Test
+    fun `given error, when build is invoked, then error data is also loaded (network)`() = test {
+        val result = mutableListOf<CardsUpdate>()
+        whenever(cardsStore.getCards(siteModel)).thenReturn(flowOf(data))
+        whenever(cardsStore.fetchCards(siteModel)).thenReturn(error)
+        cardSource.refresh.observeForever { }
+
+        cardSource.build(testScope(), SITE_LOCAL_ID).observeForever { it?.let { result.add(it) } }
+
+        assertThat(result.size).isEqualTo(1)
+        assertThat(result.first()).isEqualTo(CardsUpdate(error))
+    }
+
+    @Test
+    fun `given no error, when refresh is invoked, then data is only loaded from get cards (database)`() = test {
+        val result = mutableListOf<CardsUpdate>()
+        whenever(cardsStore.getCards(siteModel)).thenReturn(flowOf(data))
+        whenever(cardsStore.fetchCards(siteModel)).thenReturn(success).thenReturn(success)
+        cardSource.build(testScope(), SITE_LOCAL_ID).observeForever { it?.let { result.add(it) } }
+        cardSource.refresh.observeForever { }
+
+        cardSource.refresh()
+
+        assertThat(result.size).isEqualTo(1)
+        assertThat(result.first()).isEqualTo(CardsUpdate(data))
+    }
+
+    @Test
+    fun `given error, when refresh is invoked, then error data is also loaded (network)`() = test {
+        val result = mutableListOf<CardsUpdate>()
+        whenever(cardsStore.getCards(siteModel)).thenReturn(flowOf(data))
+        whenever(cardsStore.fetchCards(siteModel)).thenReturn(success).thenReturn(error)
+        cardSource.build(testScope(), SITE_LOCAL_ID).observeForever { it?.let { result.add(it) } }
+        cardSource.refresh.observeForever { }
+
+        cardSource.refresh()
+
+        assertThat(result.size).isEqualTo(2)
+        assertThat(result.first()).isEqualTo(CardsUpdate(data))
+        assertThat(result.last()).isEqualTo(CardsUpdate(error))
+    }
+
+    /* IS REFRESHING */
+
+    @Test
+    fun `when build is invoked, then refresh is set to false`() = test {
+        val result = mutableListOf<Boolean>()
+        cardSource.refresh.observeForever { result.add(it) }
+
+        cardSource.build(testScope(), SITE_LOCAL_ID).observeForever { }
+
+        assertThat(result.size).isEqualTo(1)
+        assertThat(result.last()).isFalse
+    }
+
+    @Test
+    fun `when refresh is invoked, then refresh is set to true`() = test {
+        val result = mutableListOf<Boolean>()
+        cardSource.build(testScope(), SITE_LOCAL_ID).observeForever { }
+        cardSource.refresh.observeForever { result.add(it) }
+
+        cardSource.refresh()
+
+        assertThat(result.size).isEqualTo(2)
+        assertThat(result.first()).isFalse
+        assertThat(result.last()).isTrue
+    }
+
+    @Test
+    fun `given no error, when data has been refreshed, then refresh is set to true`() = test {
```

Comment:
Shouldn't refresh be set to false after a successful fetch? If not, the isRefreshing = true would result in the refreshing indicator never being hidden?

Reply:
👋 @zwarm !

Great question, but actually no, it is a bit tricky since I couldn't find a better naming and yet group both test somehow. So, what this test wants to verify is that during a refresh, when there is no error, then nothing happens to refresh, the flow above will capture the update, thus setting the refresh to false, not fetching the cards.

However, on the test further below, the one with an error given, then here we want to verify that the refresh is set to false. It is all about mixing and matching the `build(...)` + `getCards(...)` flow, and `refresh()` + `fetchCards(...)` result.

Can you suggest a better naming for both those tests so that their intention can be understood better?
- `given no error, when data has been refreshed, then refresh is set to true`
- `given error, when data has been refreshed, then refresh is set to false`

Reply:
Updated here: 83e0684c768ce5d151a858fd8e0232f4f738888d

Maybe the update I did here will help you understand what this test is doing and why. 🤔

Reply:
I am glad it helped @zwarm but all you said are right, I just had hard time thinking of a better alternative for naming things. Feel free to suggest anything else and I can incorporate it on a subsequent commit without problem. 🙏